activerecord/test/cases/statement_cache_test.rb

```diff
@@ -105,5 +105,25 @@ def test_unprepared_statements_dont_share_a_cache_with_prepared_statements
 
       refute_equal book, other_book
     end
+
+    def test_find_by_does_not_use_statement_cache_if_table_name_is_changed
+      book = Book.create(name: "my book")
+      Book.find_by(name: "my book")
+      Book.table_name = :birds
+      assert_nil Book.find_by(name: "my book")
+    ensure
+      Book.table_name = :books
+    end
+
+    def test_find_does_not_use_statement_cache_if_table_name_is_changed
+      book = Book.create(name: "my book")
+      Book.find(book.id)
+      Book.table_name = :birds
+      assert_raise ActiveRecord::RecordNotFound do
+        Book.find(book.id)
+      end
+    ensure
+      Book.table_name = :books
+    end
```

Comment:
Nice. I think these still need some code comments explaining what's happening when we call `#find_by` and `#find` the first time (that we're warming the cache) and when we call `Book.table_name = …` (that changing the table name should change the query that's now cached).